src/smpp_gateway/subscribers.py

```diff
@@ -6,39 +6,58 @@
 from django.db import connection, transaction
 from rapidsms.router import lookup_connections, receive
 
-from smpp_gateway.models import MOMessage
+from smpp_gateway.models import MOMessage, MTMessage
 
 logger = logging.getLogger(__name__)
 
 
-def handle_mo_message(cursor):
+def handle_mo_message(notify):
     with transaction.atomic():
         sms = (
             MOMessage.objects.filter(status="new")
             .select_for_update(skip_locked=True)
             .first()
         )
-        MOMessage.objects.filter(pk=sms.pk).update(status="processing")
-    connections = lookup_connections(
-        backend="sms_gateway", identities=[sms.params["source_addr"]]
-    )
-    for conn in connections:
-        receive(sms.params["short_message"], conn)
+        if sms is not None:
+            MOMessage.objects.filter(pk=sms.pk).update(status="processing")
+    if sms is not None:
+        connections = lookup_connections(
+            backend="sms_gateway", identities=[sms.params["source_addr"]]
+        )
+        for conn in connections:
+            receive(sms.params["short_message"], conn)
 
 
-def listen_mo_messages(channel):
+def pg_listen(channel, handler):
     with connection.cursor() as cursor:
         pg_conn = connection.connection
         pg_conn.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
         # https://gist.github.com/pkese/2790749
-        cursor.execute("LISTEN new_mo_msg;")
+        cursor.execute(f"LISTEN {channel};")
         logger.info(f"Waiting for notifications on channel '{channel}'")
         while True:
             if select.select([pg_conn], [], [], 5) == ([], [], []):
-                logger.debug(".")
+                logger.debug(f"{channel} .")
             else:
                 pg_conn.poll()
                 while pg_conn.notifies:
                     notify = pg_conn.notifies.pop()
                     logger.info(f"Got NOTIFY:{notify}")
-                    handle_mo_message(cursor)
+                    handler(notify)
+
+
+def listen_mo_messages(channel):
+    pg_listen(channel, handle_mo_message)
+
+
+def get_mt_messages(channel, limit):
```

Comment:
This might be better in `MTMessage`... something to fetch messages and update their status atomically.